File: packages/app/src/services/software/definitions/librewolf.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'librewolf';
	public name = 'LibreWolf';
	public category = [SoftwareCategory.Browser];
	public downloadName = 'librewolf-windows-x86_64-setup.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public deprecated = false;
	public icon = 'librewolf.png';
	public homepage = 'https://librewolf.net';

	public async resolveDownloadUrl() {
		const res = await fetch('https://librewolf.net/installation/windows');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/"(https:\/\/gitlab.com\/api\/v4\/projects\/\d{8}\/packages\/generic\/librewolf\/.*\/librewolf-.*-windows-x86_64-setup\.exe)"/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}
}

File: src/lib/software/librewolf.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class LibreWolf(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'librewolf'
        self.name = 'LibreWolf'
        self.category = [SoftwareCategory.Browser]
        self.download_name = 'librewolf-windows-x86_64-setup.exe'
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'librewolf.png'
        self.homepage = 'https://librewolf.net'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'"(https://gitlab.com/api/v4/projects/\d{8}/packages/generic/librewolf/.*/librewolf-.*-windows-x86_64-setup\.exe)"')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://librewolf.net/installation/windows/'))
